File: examples/Bayesian_estimation_singleparameter.py
from quanestimation import *
import numpy as np
from scipy.integrate import simps
import random

# initial state
rho0 = 0.5 * np.array([[1.0, 1.0], [1.0, 1.0]])
# free Hamiltonian
B = np.pi / 2.0
x = [np.linspace(0.0, 0.5 * np.pi, 1000)]
sx = np.array([[0.0, 1.0], [1.0, 0.0]])
sy = np.array([[0.0, -1.0j], [1.0j, 0.0]])
sz = np.array([[1.0, 0.0], [0.0, -1.0]])
H0_func = lambda x: 0.5 * B * (sx * np.cos(x) + sz * np.sin(x))
dH_func = lambda x: [0.5 * B * (-sx * np.sin(x) + sz * np.cos(x))]
# measurement
M1 = 0.5 * np.array([[1.0, 1.0], [1.0, 1.0]])
M2 = 0.5 * np.array([[1.0, -1.0], [-1.0, 1.0]])
M = [M1, M2]
# dynamics
tspan = np.linspace(0.0, 1.0, 1000)
# prior distribution
x = [np.linspace(0.0, 0.5 * np.pi, 100)]
p = (1.0 / (x[0][-1] - x[0][0])) * np.ones(len(x[0]))

rho = [[] for i in range(len(x[0]))]
for i in range(len(x[0])):
    H0_tp = H0_func(x[0][i])
    dH_tp = dH_func(x[0][i])
    dynamics = Lindblad(tspan, rho0, H0_tp, dH_tp)
    rho_tp, drho_tp = dynamics.expm()
    rho[i] = rho_tp[-1]
# generate experimental results
np.random.seed(1234)
y = [0 for i in range(500)]
res_rand = random.sample(range(0, len(y)), 125)
for i in range(len(res_rand)):
    y[res_rand[i]] = 1

pout, xout = Bayes(x, p, rho, y, M=M, savefile=False)
print(xout)
Lout, xout = MLE(x, rho, y, M=M, savefile=False)
print(xout)

File: quanestimation/examples/Julia examples/Bayesian_estimation_singleparameter.jl
using Random
using StatsBase
include("../src/QuanEstimation.jl")

# initial state
rho0 = 0.5*[1.0 1.0+0.0im; 1.0 1.0]
# Hamiltonian
B = pi/2.0
sx = [0.0 1.0; 1.0 0.0im]
sy = [0.0 -1.0im; 1.0im 0.0]
sz = [1.0 0.0im; 0.0 -1.0]
function H0_func(x)
    return 0.5*B*(sx*cos(x)+sz*sin(x))
end
function dH_func(x)
    return [0.5*B*(-sx*sin(x)+sz*cos(x))]
end
# measurement 
M1 = 0.5*[1.0+0.0im  1.0; 1.0  1.0]
M2 = 0.5*[1.0+0.0im -1.0; -1.0  1.0]
M = [M1, M2]
# dynamics
decay_opt = [zeros(ComplexF64,size(rho0)[1],size(rho0)[1])] 
gamma = [0.0]
tspan = range(0.0, stop=1.0, length=1000)

#### prior distribution ####
x = [range(0.0, stop=pi/2.0, length=100)].|>Vector
p = (1.0/(x[1][end]-x[1][1]))*ones(length(x[1]))
dp = zeros(length(x[1]))

rho = Vector{Matrix{ComplexF64}}(undef, length(x[1]))
for i = 1:length(x[1]) 
    H0_tp = H0_func(x[1][i])
    dH_tp = dH_func(x[1][i])
    rho_tp, drho_tp  = QuanEstimation.expm(H0_tp, dH_tp, [zeros(ComplexF64,size(rho0)[1],size(rho0)[1])], [zeros(length(tspan)-1)], rho0, tspan, decay_opt, gamma)
    rho[i] = rho_tp[end]
end

# generate experimental results
Random.seed!(1234)
y = [0 for i in 1:500]
res_rand = sample(1:length(y), 125, replace=false)
for i in 1:length(res_rand)
    y[res_rand[i]] = 1
end

pout, xout = QuanEstimation.Bayes(x, p, rho, y; M=M, savefile=false)
println(xout)
Lout, xout = QuanEstimation.MLE(x, rho, y, M=M; savefile=false)
println(xout)
